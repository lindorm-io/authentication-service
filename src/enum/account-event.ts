export enum AccountEvent {
  CREATED = "account_created",
  EMAIL_CHANGED = "account_email_changed",
  OTP_CHANGED = "account_otp_changed",
  PASSWORD_CHANGED = "account_password_changed",
  PERMISSION_CHANGED = "account_permission_changed",
}
