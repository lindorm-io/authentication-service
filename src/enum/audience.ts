export enum Audience {
  ACCESS = "access",
  AUTHORIZATION = "authorization",
  IDENTITY = "identity",
  MULTI_FACTOR = "multi-factor",
  REFRESH = "refresh",
}
