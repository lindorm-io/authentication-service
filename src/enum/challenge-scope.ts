export enum ChallengeScope {
  CHANGE = "change",
  SIGN_IN = "sign_in",
  RECOVERY = "recovery",
}
