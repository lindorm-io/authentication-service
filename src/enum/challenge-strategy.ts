export enum ChallengeStrategy {
  IMPLICIT = "implicit",
  PIN = "pin",
  SECRET = "secret",
}
