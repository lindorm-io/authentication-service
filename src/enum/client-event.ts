export enum ClientEvent {
  CREATED = "client_created",
  APPROVED_CHANGED = "client_approved_changed",
  DESCRIPTION_CHANGED = "client_description_changed",
  EMAIL_AUTHORIZATION_URI_CHANGED = "client_email_authorization_uri_changed",
  NAME_CHANGED = "client_name_changed",
  SECRET_CHANGED = "client_secret_changed",
}
