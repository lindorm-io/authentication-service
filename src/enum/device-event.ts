export enum DeviceEvent {
  CREATED = "device_created",
  PIN_CHANGED = "device_pin_changed",
  SECRET_CHANGED = "device_secret_changed",
  NAME_CHANGED = "device_name_changed",
}
