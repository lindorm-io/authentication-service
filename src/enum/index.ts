export * from "./account-event";
export * from "./audience";
export * from "./device-event";
export * from "./grant-type";
export * from "./mongo-collection";
export * from "./multi-factor-challenge-type";
export * from "./response-type";
export * from "./session-event";
