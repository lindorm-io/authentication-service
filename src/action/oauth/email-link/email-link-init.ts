import Joi from "@hapi/joi";
import { GrantType } from "../../../enum";
import { IKoaAuthContext } from "../../../typing";
import { JOI_CODE_CHALLENGE, JOI_CODE_METHOD, JOI_EMAIL, JOI_GRANT_TYPE, JOI_STATE } from "../../../constant";
import { Scope } from "@lindorm-io/jwt";
import { assertValidScopeInput, assertValidResponseTypeInput } from "../../../util";
import { createAuthorization, getAuthorizationToken, sendEmailLink } from "../../../support";

export interface IPerformEmailLinkInitOptions {
  codeChallenge: string;
  codeMethod: string;
  grantType: string;
  redirectUri: string;
  responseType: string;
  scope: string;
  state: string;
  subject: string;
}

export interface IPerformEmailLinkInitData {
  expires: number;
  expiresIn: number;
  state: string;
}

const schema = Joi.object({
  codeChallenge: JOI_CODE_CHALLENGE,
  codeMethod: JOI_CODE_METHOD,
  grantType: JOI_GRANT_TYPE,
  redirectUri: Joi.string().uri().required(),
  responseType: Joi.string().required(),
  scope: Joi.string().required(),
  state: JOI_STATE,
  subject: JOI_EMAIL,
});

export const performEmailLinkInit = (ctx: IKoaAuthContext) => async (
  options: IPerformEmailLinkInitOptions,
): Promise<IPerformEmailLinkInitData> => {
  await schema.validateAsync(options);

  const { client } = ctx;
  const { codeChallenge, codeMethod, grantType, redirectUri, responseType, state, subject } = options;
  const scope = options.scope.split(" ") as Array<Scope>;

  assertValidResponseTypeInput(responseType);
  assertValidScopeInput(scope);

  const authorization = await createAuthorization(ctx)({
    codeChallenge,
    codeMethod,
    email: subject,
    grantType: GrantType.EMAIL_LINK,
    redirectUri,
    responseType,
    scope,
  });

  const { expires, expiresIn, token } = getAuthorizationToken(ctx)({ authorization, client });

  await sendEmailLink(ctx)({
    grantType,
    redirectUri,
    state,
    subject,
    token,
  });

  return {
    expires,
    expiresIn,
    state,
  };
};
