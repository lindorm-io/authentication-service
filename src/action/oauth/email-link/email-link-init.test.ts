import { GrantType, ResponseType } from "../../../enum";
import { Scope } from "@lindorm-io/jwt";
import { performEmailLinkInit } from "./email-link-init";

jest.mock("../../../util", () => ({
  assertResponseType: jest.fn(() => undefined),
}));
jest.mock("../../../support", () => ({
  createSession: jest.fn(() => () => "session"),
  getAuthorizationToken: jest.fn(() => () => ({
    expires: "expires",
    expiresIn: "expiresIn",
    token: "token",
  })),
  sendEmailLink: jest.fn(() => () => undefined),
}));

describe("performEmailLinkInit", () => {
  let getMockContext: any;

  beforeEach(() => {
    getMockContext = () => ({
      client: "client",
    });
  });

  test("should create a new session", async () => {
    await expect(
      performEmailLinkInit(getMockContext())({
        codeChallenge: "Z1teIWMlf6xFacp4quXP3O0XI204ZT1b",
        codeMethod: "sha512",
        grantType: GrantType.DEVICE_PIN,
        redirectUri: "https://redirect.uri/",
        responseType: ResponseType.REFRESH,
        scope: Scope.DEFAULT,
        state: "1bVcJqZ1pBeqVLxV",
        subject: "email@lindorm.io",
      }),
    ).resolves.toStrictEqual({
      expires: "expires",
      expiresIn: "expiresIn",
      state: "1bVcJqZ1pBeqVLxV",
    });
  });
});
