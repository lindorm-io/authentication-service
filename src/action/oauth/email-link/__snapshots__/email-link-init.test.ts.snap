// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`performEmailLinkInit should create a new session 1`] = `
Object {
  "expires": "expires",
  "expiresIn": "expiresIn",
  "state": "1bVcJqZ1pBeqVLxV",
}
`;
